File: OpenCount/Models/CountRegion.swift
import Foundation
import CoreGraphics
import SwiftData

/// The shape type for a counting region.
enum RegionShapeType: String, Codable {
    case rectangle
    case ellipse
    case polygon
}

/// A user-drawn geometric area on an image that restricts counting to objects within its boundary.
/// All coordinates are normalized (0.0–1.0) relative to the image dimensions.
@Model
final class CountRegion {
    var id: UUID
    var name: String
    /// Hex color string, e.g. "#3399FF"
    var colorHex: String
    var shapeType: RegionShapeType
    /// Polygon vertices or rect corners in normalized coordinates
    var normalizedPoints: [CGPoint]
    var session: CountSession?

    init(
        id: UUID = UUID(),
        name: String,
        colorHex: String = "#3399FF",
        shapeType: RegionShapeType = .rectangle,
        normalizedPoints: [CGPoint] = [],
        session: CountSession? = nil
    ) {
        self.id = id
        self.name = name
        self.colorHex = colorHex
        self.shapeType = shapeType
        self.normalizedPoints = normalizedPoints
        self.session = session
    }

    // MARK: - Geometry helpers

    /// Returns true if the given normalized point falls within this region's boundary.
    func contains(normalizedPoint point: CGPoint) -> Bool {
        switch shapeType {
        case .rectangle:
            return containsRectangle(point)
        case .ellipse:
            return containsEllipse(point)
        case .polygon:
            return containsPolygon(point)
        }
    }

    private func containsRectangle(_ point: CGPoint) -> Bool {
        guard normalizedPoints.count >= 2 else { return false }
        let minX = normalizedPoints.map(\.x).min() ?? 0
        let maxX = normalizedPoints.map(\.x).max() ?? 0
        let minY = normalizedPoints.map(\.y).min() ?? 0
        let maxY = normalizedPoints.map(\.y).max() ?? 0
        return point.x > minX && point.x < maxX && point.y > minY && point.y < maxY
    }

    private func containsEllipse(_ point: CGPoint) -> Bool {
        guard normalizedPoints.count >= 2 else { return false }
        let minX = normalizedPoints.map(\.x).min() ?? 0
        let maxX = normalizedPoints.map(\.x).max() ?? 0
        let minY = normalizedPoints.map(\.y).min() ?? 0
        let maxY = normalizedPoints.map(\.y).max() ?? 0
        let cx = (minX + maxX) / 2
        let cy = (minY + maxY) / 2
        let rx = (maxX - minX) / 2
        let ry = (maxY - minY) / 2
        guard rx > 0, ry > 0 else { return false }
        let dx = (point.x - cx) / rx
        let dy = (point.y - cy) / ry
        return (dx * dx + dy * dy) < 1.0
    }

    private func containsPolygon(_ point: CGPoint) -> Bool {
        guard normalizedPoints.count >= 3 else { return false }
        // Ray-casting algorithm
        var inside = false
        var j = normalizedPoints.count - 1
        for i in 0..<normalizedPoints.count {
            let pi = normalizedPoints[i]
            let pj = normalizedPoints[j]
            let intersects = ((pi.y > point.y) != (pj.y > point.y)) &&
                (point.x < (pj.x - pi.x) * (point.y - pi.y) / (pj.y - pi.y) + pi.x)
            if intersects { inside.toggle() }
            j = i
        }
        return inside
    }
}

File: packages/shared/src/commonMain/kotlin/com/opencount/shared/model/CountRegion.kt
package com.opencount.shared.model

import kotlinx.serialization.Serializable

@Serializable
enum class RegionShapeType {
    Rectangle,
    Ellipse,
    Polygon,
}

@Serializable
data class NormalizedPoint(
    val x: Double,
    val y: Double,
)

@Serializable
data class CountRegion(
    val id: String,
    val name: String,
    val colorHex: String = "#3399FF",
    val shapeType: RegionShapeType = RegionShapeType.Rectangle,
    val normalizedPoints: List<NormalizedPoint> = emptyList(),
) {
    fun contains(normalizedPoint: NormalizedPoint): Boolean = when (shapeType) {
        RegionShapeType.Rectangle -> containsRectangle(normalizedPoint)
        RegionShapeType.Ellipse -> containsEllipse(normalizedPoint)
        RegionShapeType.Polygon -> containsPolygon(normalizedPoint)
    }

    private fun containsRectangle(point: NormalizedPoint): Boolean {
        if (normalizedPoints.size < 2) return false
        val xs = normalizedPoints.map { it.x }
        val ys = normalizedPoints.map { it.y }
        return point.x > (xs.minOrNull() ?: 0.0) && point.x < (xs.maxOrNull() ?: 0.0)
            && point.y > (ys.minOrNull() ?: 0.0) && point.y < (ys.maxOrNull() ?: 0.0)
    }

    private fun containsEllipse(point: NormalizedPoint): Boolean {
        if (normalizedPoints.size < 2) return false
        val xs = normalizedPoints.map { it.x }
        val ys = normalizedPoints.map { it.y }
        val cx = ((xs.minOrNull() ?: 0.0) + (xs.maxOrNull() ?: 0.0)) / 2.0
        val cy = ((ys.minOrNull() ?: 0.0) + (ys.maxOrNull() ?: 0.0)) / 2.0
        val rx = ((xs.maxOrNull() ?: 0.0) - (xs.minOrNull() ?: 0.0)) / 2.0
        val ry = ((ys.maxOrNull() ?: 0.0) - (ys.minOrNull() ?: 0.0)) / 2.0
        if (rx <= 0.0 || ry <= 0.0) return false
        val dx = (point.x - cx) / rx
        val dy = (point.y - cy) / ry
        return (dx * dx + dy * dy) < 1.0
    }

    private fun containsPolygon(point: NormalizedPoint): Boolean {
        if (normalizedPoints.size < 3) return false
        var inside = false
        var j = normalizedPoints.size - 1
        for (i in normalizedPoints.indices) {
            val pi = normalizedPoints[i]
            val pj = normalizedPoints[j]
            if ((pi.y > point.y) != (pj.y > point.y) &&
                point.x < (pj.x - pi.x) * (point.y - pi.y) / (pj.y - pi.y) + pi.x
            ) {
                inside = !inside
            }
            j = i
        }
        return inside
    }
}
